### PR TITLE
Fail earlier on cache corruptions

### PR DESCRIPTION
**Pull request by yschimke**
Fail earlier on cache corruption, log warning instead of failing terminally.

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/HttpUrl.kt [hunk 1]
```diff
@@ -1258,7 +1258,7 @@ class HttpUrl internal constructor(
         this.scheme = base.scheme
       } else {
         throw IllegalArgumentException(
-            "Expected URL scheme 'http' or 'https' but no colon was found")
+            "Expected URL scheme 'http' or 'https' but no scheme was found for $input")
```
Comment (yschimke) [inline okhttp/src/main/kotlin/okhttp3/HttpUrl.kt:1261]: Happy to drop this for PII reasons if there is concern, although as a corrupted url, the debug value seems higher than normal.

File: okhttp/src/main/kotlin/okhttp3/Cache.kt [hunk 2]
```diff
@@ -490,9 +500,14 @@ class Cache internal constructor(
...
      */
     @Throws(IOException::class) constructor(rawSource: Source) {
-      try {
+      rawSource.use {
         val source = rawSource.buffer()
-        url = source.readUtf8LineStrict()
+        val urlLine = source.readUtf8LineStrict()
+        // Choice here is between failing with a correct RuntimeException
+        // or mostly silently with an IOException
+        url = urlLine.toHttpUrlOrNull() ?: throw IOException("Cache Failure for $urlLine").also {
```
Comment (yschimke) [inline okhttp/src/main/kotlin/okhttp3/Cache.kt:508]: 3 options here

1) runtime exception - fatal...
2) io exception only
3) io exception and log a warning

File: okhttp/src/main/kotlin/okhttp3/HttpUrl.kt (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/kotlin/okhttp3/HttpUrl.kt:1261]: Log up to 6 characters? Like `www.ex...` ?

File: okhttp/src/main/kotlin/okhttp3/Cache.kt (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/kotlin/okhttp3/Cache.kt:508]: change `cache failure` to `cache corruption` ?

**Review by swankjesse — APPROVED** [review]

**Comment by Thomas-Vos** [comment]
Could this fix be backported to 4.9.x, please?

**Pull request merged**